Add configuration for cross-account AWS database-access

GavinFrazar opened the pull request:
I am splitting up PRs for the implementation of https://github.com/gravitational/teleport/issues/21872 to reduce PR size.

This PR adds configuration for AWS `assume_role_arn` in database config and discovery matchers.
It does not add the logic that actually uses this new configuration - that will be added in PRs to follow.

This new config is available in:
- `teleport db configure create`
- `teleport db start`
  - I know we want to avoid adding flags to this command because we don't recommend it in our docs, but `--aws-external-id` was already a flag and `--aws-assume-role-arn` naturally pairs with that flag.
- `tctl create/edit`
- yaml config file loaded by `teleport`

merge base is set to the branch for https://github.com/gravitational/teleport/pull/23157 which has various changes I split out for backporting and will update to master once that branch merges.

GavinFrazar committed:
Add configuration for AWS assume_role_arn

GavinFrazar committed:
improve aws utils and database validation

* log a warning when an invalid AWS IAM role ARN is filtered out
* add tests for nested role arn path and remove superfluous test parallelism
* add extra godoc info for BuildRoleARN
* improve mock IAM error messages
  - Failing tests' error messages were not specific enough to know what
    happened.
* refactor code to call GetAWS() once
  - Avoids constantly fetching a full copy of the AWS metadata.
  - Reduces verbosity of code.
* reuse MetadataFromRedshiftCluster in meta fetcher
  - Avoids repeating the same logic inside the fetcher.
* improve database-access DynamoDB config checking
  - Don't early return from database CheckAndSetDefaults so AWS account ID
    is checked properly, and any other future checks added will apply to
    DynamoDB as well.
* add aws utils for parsing AWS IAM roles
  - ParseRoleARN will parse an AWS ARN and verify that it is a valid
    AWS IAM Role resource
  - CheckARNPartitionAndAccount is a helper that verifies an ARN partition
    and account ID matches the caller's expectations.
* use cmp.Diff instead of require.Equal in db tests
  - require.Equal spams a lot of protobuf noise.
  - cmp.Diff yields a precise, readable diff on test failure.

greedy52 commented on api/types/database_test.go:
```diff
@@ -564,6 +566,22 @@ func TestDynamoDBConfig(t *testing.T) {
+				AWS: AWS{
+					Region:        "us-west-1",
+					AccountID:     "123456789012",
+					AssumeRoleARN: "arn:aws:iam::123456789012:role/DBDiscoverer",
```
Curious why `AssumeRoleARN` is required in this case? I could be wrong but I feel `tsh db connect --db-user arn:aws:iam::<another-account>:role/dynamodb-role my-dynamodb` is legit with `ExternalID` configured?

GavinFrazar replied:
Good point. We have 3 protocols (cassandra, dynamodb, redshift serverless) that assume a role for authentication, and the role is built from `--db-user`.

Of those 3, only dynamodb is passing `external_id` when it assumes the role.

I discussed with @smallinsky that for these protocols we might have two roles to assume but only one external id - so which role should pair with the external ID?

We decided it would be be best to pair external ID with the configured `assume_role_arn`, so that cross-account access delegation only needs to be setup for one role rather than multiple.

The problem with that change is DynamoDB, which is already supporting `external_id`...

I suppose I could remove this check for cassandra/redshiftserverless/dynamodb specifically and just pass `external_id` with both assume role calls, since it won't have any effect for roles that are already in the same account as the caller. That would prevent breaking existing config for anyone using external ID w/ DynamoDB, and make the behavior consistent for all 3 protocols. What do you guys think? Any issues with doing that?

GavinFrazar replied:
An alternative solution is to check the database's configured `AccountID` and only pass external id when assuming a configured `assume_role_arn` if the account ID doesn't match the role account ID, and likewise for assuming a role from `--db-user`. That might be a cleaner solution.

greedy52 replied:
hmm. chained assume role? that's interesting. Would it make sense to have two external IDs for each? not sure if this is more confusing but something like this:

host role -> assume "discovery role" with `aws.AssumeRole.ARN` and `aws.AssumeRole.ExternalID` -> assume db user role with `--db-user` and `aws.ExternalID`

greedy52 replied:
> I suppose I could remove this check for cassandra/redshiftserverless/dynamodb specifically and just pass external_id with both assume role calls, since it won't have any effect for roles that are already in the same account as the caller. That would prevent breaking existing config for anyone using external ID w/ DynamoDB, and make the behavior consistent for all 3 protocols. What do you guys think? Any issues with doing that?

Or, only use external_id for "discovery role" if it's set, and if "discovery role" not set, use it for the "db user role"?

GavinFrazar replied:
> host role -> assume "discovery role" with aws.AssumeRole.ARN and aws.AssumeRole.ExternalID -> assume db user role with --db-user and aws.ExternalID

> only use external_id for "discovery role" if it's set, and if "discovery role" not set, use it for the "db user role"?

I think it might be too confusing to have two configurable external IDs. I prefer the second suggestion. For config validation, that means I'll have a special exception that dynamodb/redshift serverless/cassandra will not require `assume_role_arn` with `external_id`.

GavinFrazar replied:
And to add onto this, I think if `assume_role_arn` is given, no matter the db protocol, we should require it matches `account_id` (or if `account_id` is blank, set it to match `assume_role_arn`).

Reasoning is as follows:
 
Technically for fetching an auth token or signing requests, `--db-user` could provide a role in the same account as the database. However, for metadata fetching and IAM configuration, the agent will only know about `assume_role_arn`, and those operations will fail if the database is in a different account than the role that the agent assumes.

Example:
```yaml
db_service:
  enabled: true
  databases:
    - name: "ddb"
      protocol: "dynamodb"
      static_labels:
        foo: "bar"
      aws:
        region: "us-west-1"
        account_id: "123456789012"
        assume_role_arn: "arn:aws:iam::278576220453:role/role-in-external-account"
        external_id: "1234"
```

```
# works for proxying requests, but agent IAM configurator is failing since it has an external `assume_role_arn`
tsh proxy db --tunnel --db-user="arn:aws:iam::123456789012:role/role-in-same-account" ddb
```

If `assume_role_arn` is instead left blank, then the agent IAM configurator will use `sts get-caller-identity` to configure itself.
If `assume_role_arn` is not blank but matches the database account ID, then agent IAM configurator will work in an external account.

GavinFrazar replied:
note that in the case where `--db-user` is an external account and `assume_role_arn` is blank, the agent's configurator won't be able to configure IAM across accounts because it doesn't know about potential settings for `--db-user`. This limitation means we should recommend in docs that for cross account AWS, they should use `assume_role_arn` with `external_id` instead of supplying the external role via `--db-user`.

We do not currently have IAM configurator for DynamoDB, so existing customers are having to configure IAM manually already anyway. I plan to add that as well. I don't think this will regress anything.

GavinFrazar replied:
I made the change for `CheckAndSetDefaults`. It only errors if `d.RequireAWSIAMRolesAsUsers()` returns false. This covers keyspaces, redshift serverless, and dynamodb.

greedy52 commented on lib/service/servicecfg/database.go:
```diff
@@ -99,6 +100,17 @@ func (d *Database) CheckAndSetDefaults() error {
 		}
 	}
 
+	// if AWS account id is missing, but assume role arn is given,
```
nit
```suggestion
	// If AWS account ID is missing, but assume role ARN is given,
```

GavinFrazar committed:
Merge branch 'master' into gavinfrazar/support-cross-account-aws-discovery-config

GavinFrazar committed:
Update lib/service/servicecfg/database.go

Co-authored-by: STeve (Xin) Huang <xin.huang@goteleport.com>

GavinFrazar committed:
fixup database checks

* external_id does not require assume_role_arn
  for DynamoDB, Keyspaces, or Redshift Serverless
* keyspaces gives a hint that URI could not be
  derived because region was also empty
* make some error messages more consistent with
  each other regarding missing/empty fields.

greedy52 commented on lib/service/servicecfg/database.go:
```diff
@@ -99,6 +100,17 @@ func (d *Database) CheckAndSetDefaults() error {
 		}
 	}
 
+	// If AWS account ID is missing, but assume role ARN is given,
+	// try to parse the role arn and set the account id to match.
+	if d.AWS.AccountID == "" && d.AWS.AssumeRoleARN != "" {
+		parsed, err := awsutils.ParseRoleARN(d.AWS.AssumeRoleARN)
+		if err != nil {
+			return trace.BadParameter("database %q invalid AWS assume_role_arn: %v",
+				d.Name, err)
+		}
+		d.AWS.AccountID = parsed.AccountID
+	}
+
```
nit: Not a problem started by this change but ideally we should do this in api.Database's `CheckAndSetDefaults` so that databases created by tctl and webapi can benefit from it. (maybe the same for the TLS and AD `CheckAndSetDefaults` above, if we can move them to api if possible)

@smallinsky @GavinFrazar do you think if worth copying the AWS arn implementation into our api lib?
https://github.com/aws/aws-sdk-go/blob/v1.44.224/aws/arn/arn.go
It's not that much code. And if we have our own implementation, we can even enhance it with more functions we need.

GavinFrazar replied:
👍  Yeah I put it here to avoid pulling AWS sdk into our api. I have the same check in `ValidateDatabase`, but it just errors for empty AccountID because we don't set defaults in that function. Would definitely be a better UX for tctl and webapi if we can move these all into the api `CheckAndSetDefaults`.

smallinsky replied:
> do you think if worth copying the AWS arn implementation into our api lib?
https://github.com/aws/aws-sdk-go/blob/v1.44.224/aws/arn/arn.go

Yea we can do that but why not just embed the aws.ARN into own custom type and add additional logic.

greedy52 replied:
> Yea we can do that but why not just embed the aws.ARN into own custom type and add additional logic ?

I think we hesitate whether to bring the whole AWS SDK dependency to api's go.mod. 

Another alternative is to get `aws-sdk-go-v2` without other services.

GavinFrazar replied:
I think I want to just pull in `aws-sdk-go-v2` since the stripped binary size change is tiny and v2 doesn't download the whole sdk when fetching dependencies, but I should move this code in another PR to get feedback from other people first. It enables only a small UX improvement for `tctl` and web api. I can move the TLS and AD stuff too in that PR.

GavinFrazar committed:
Merge branch 'master' into gavinfrazar/support-cross-account-aws-discovery-config

GavinFrazar committed:
Merge branch 'master' into gavinfrazar/support-cross-account-aws-discovery-config
